plugins/parsers/csv/parser.go

```diff
@@ -117,15 +214,41 @@ func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
 }
 
 func parseCSV(p *Parser, r io.Reader) ([]telegraf.Metric, error) {
-	csvReader := p.compile(r)
+	lineReader := bufio.NewReader(r)
 	// skip first rows
 	for p.SkipRows > 0 {
-		_, err := csvReader.Read()
-		if err != nil {
+		line, err := lineReader.ReadString('\n')
+		if err != nil && len(line) == 0 {
 			return nil, err
 		}
 		p.SkipRows--
 	}
+	// Parse metadata
+outer:
+	for p.MetadataRows > 0 {
+		line, err := lineReader.ReadString('\n')
+		if err != nil && len(line) == 0 {
+			return nil, err
+		}
+		p.MetadataRows--
+		m := p.ParseMetadataRow(line)
+		if m != nil {
+			for _, tagName := range p.TagColumns {
+				if tagName == m.key {
+					if p.DefaultTags == nil {
+						p.DefaultTags = make(map[string]string)
+					}
+					p.DefaultTags[m.key] = m.value
+					continue outer
+				}
+			}
+			if p.DefaultFields == nil {
+				p.DefaultFields = make(map[string]string)
+			}
+			p.DefaultFields[m.key] = m.value
+		}
```

Comment:
TBH I don't get the meaning of this. Are you trying to decide of the found meta-data is a tag or field? If so, please don't. You can still convert the between them afterwards, using a processor. Please also avoid messing with `DefaultTags` as those are set by telegraf and you should not make any assumption whatsoever on when it is set.